File: Java/src/main/java/org/example/Rectangle.java
package org.example;

public class Rectangle {
    public float xPos, yPos;
    public float width, height;
    public float r, g, b;
    private float xVel, yVel;
    private boolean collided;

    public Rectangle(float xPos, float yPos, float width, float height, float r, float g, float b, float xVel, float yVel) {
        this.xPos = xPos;
        this.yPos = yPos;
        this.width = width;
        this.height = height;
        this.r = r;
        this.b = b;
        this.g = g;
        this.xVel = xVel;
        this.yVel = yVel;
        collided = false;
    }

    public void updatePosition(float deltaTime) {
        xPos += xVel * deltaTime;
        yPos += yVel * deltaTime;
    }

    public void checkBoundingBox(int WINDOW_WIDTH, int WINDOW_HEIGHT) {
        if (xPos < 0) {
            xPos = 0;
            xVel = -xVel;
            collided = true;
        } else if (xPos + width > WINDOW_WIDTH) {
            xPos = WINDOW_WIDTH - width;
            xVel = -xVel;
            collided = true;
        }
        if (yPos < 0) {
            yPos = 0;
            yVel = -yVel;
            collided = true;
        } else if (yPos + height > WINDOW_HEIGHT) {
            yPos = WINDOW_HEIGHT - height;
            yVel = -yVel;
            collided = true;
        }
    }

    public void handleCollision() {
        if (collided) {
            r = (float) Math.random();
            g = (float) Math.random();
            b = (float) Math.random();
            collided = false;
        }
    }
}

File: Cpp/src/Rectangle.cpp
#include "Rectangle.h"

namespace exj {
    Rectangle::Rectangle() {

    }

    Rectangle::Rectangle(float xPos, float yPos, float width, float height, float xVel, float yVel, bool collided, float r,
                         float g, float b) : xPos_(xPos), yPos_(yPos), width_(width), height_(height), xVel_(xVel),
                                       yVel_(yVel), collided(collided), r(r), g(g), b(b) {}

    Rectangle::~Rectangle() {

    }

    void Rectangle::move(float dt) {
        xPos_ += xVel_ * dt;
        yPos_ += yVel_ * dt;
    }

    void Rectangle::checkCollisions(float width, float height) {
        if (xPos_ < 0) {
            xPos_ = 0;
            xVel_ = -xVel_;
            collided = true;
        } else if (xPos_ > width - width_) {
            xPos_ = width - width_;
            xVel_ = -xVel_;
            collided = true;
        }
        if (yPos_ < 0) {
            yPos_ = 0;
            yVel_ = -yVel_;
            collided = true;
        } else if (yPos_ > height - height_) {
            yPos_ = height - height_;
            yVel_ = -yVel_;
            collided = true;
        }
    }

    void Rectangle::handleCollisions() {
        if (collided) {
            collided = false;
            r = (float) rand()/RAND_MAX;
            g = (float) rand()/RAND_MAX;
            b = (float) rand()/RAND_MAX;
        }
    }


} // exj